### PR TITLE
Cloud 352

### PR DESCRIPTION
**Pull request by das-vinculum**
@rebuy-de/prp-kubernetes-deployment Beware, we introduced templating. Please review.

### CODE REVIEW

File: app.go [hunk 1]
```diff
@@ -148,7 +189,34 @@ func (app *App) FetchService(service *Service) error {
 			return err
 		}
 	}
+	return nil
+}
+
+func (app *App) RenderTemplates(config *ProjectConfig) error {
```
Comment (das-vinculum) [inline app.go]: Make a switch to not rerender the templates.

File: app.go [hunk 2]
```diff
@@ -45,27 +51,64 @@ func (app *App) Run() error {
...
+	}
+
+	app.OutputPath = *config.Settings.Output
+	app.SleepInterval = *config.Settings.Sleep
+	app.IgnoreDeployFailures = *config.Settings.IgnoreDeployFailures
+	app.RetrySleep = *config.Settings.RetrySleep
+	app.RetryCount = *config.Settings.RetryCount
+	app.SkipShuffle = *config.Settings.SkipShuffle
+	app.SkipFetch = *config.Settings.SkipFetch
+	app.SkipDeploy = *config.Settings.SkipDeploy
```
Comment (das-vinculum) [inline app.go]: `Skip*` should stay command line parameters

File: app.go [hunk 3]
```diff
@@ -185,29 +251,33 @@ func (app *App) DeployService(service *Service) error {
 			service.Name, manifestPath)
 	}
 
-	for _, manifest := range manifests {
-		log.Printf("Applying manifest '%s'", manifest)
+	for _, manifestInputFile := range manifests {
```
Comment (das-vinculum) [inline app.go]: `%s/manifestInputFile/manifestFile/g`

File: app.go [hunk 4]
```diff
@@ -185,29 +251,33 @@ func (app *App) DeployService(service *Service) error {
...
 		}
 	}
-
 	return nil
 }
 
-func (app *App) DisplayErrors() {
+func (app *App) handleErrors() {
 	if len(app.Errors) < 1 {
+		os.RemoveAll(app.WorkPath)
```
Comment (das-vinculum) [inline app.go]: We do not have the need for `app.WorkPath`anymore.

File: config/services.yaml [hunk 5]
```diff
@@ -1,6 +1,20 @@
+settings:
+  kubeconfig: ~/config/kubeconfig.yaml
+  output: ./output
+  sleep: 1s
+  skip-shuffle: false
+  skip-fetch: false
+  skip-deploy: false
```
Comment (das-vinculum) [inline config/services.yaml]: `skip-*` shouldn't be in here.

File: config/services.yaml [hunk 6]
```diff
@@ -1,6 +1,20 @@
...
+  retry-sleep: 250ms
+  retry-count: 3
+  ignore-deploy-failures: false
```
Comment (das-vinculum) [inline config/services.yaml]: Maybe this should also be a command line parameter.

File: hack/build.sh [hunk 7]
```diff
@@ -1,13 +1,12 @@
...
-ROOT=$(readlink -f $( dirname $0)/..)
+ROOT=$(dirname $(dirname $0))
 COMMAND=${1:-all}
 
-set -ex
-
 docker run \
 	--rm \
-	-v "${ROOT}:/go/src/${REPO}" \
+	-v "${PWD}:/go/src/${REPO}" \
```
Comment (das-vinculum) [inline hack/build.sh]: use `ROOT`again

File: templates/parse_manifests.go [hunk 8]
```diff
@@ -0,0 +1,28 @@
+package templates
+
+import (
+	"text/template"
+	"os"
+)
+
+func ParseManifestFile(inputFile string, outputFile string, settings map[string]string) error {
```
Comment (das-vinculum) [inline templates/parse_manifests.go]: Use this!

File: app.go [hunk 9]
```diff
@@ -2,20 +2,23 @@ package main
...
-	"github.com/rebuy-de/kubernetes-deployment/git"
-	"github.com/rebuy-de/kubernetes-deployment/kubernetes"
 )
 
 type App struct {
 	Kubectl           kubernetes.API
 	ProjectConfigPath string
+	LocalConfigPath   string
 	OutputPath        string
+	WorkPath          string
```
Comment (svenwltr) [inline app.go]: If I understood it correctly, this isn't needed anymore.

File: app.go (hunk 1 above)
Comment (svenwltr) [inline app.go]: Maybe split this also into `RenderTemplates` and `RenderTemplate` to make this better readable.

File: app.go [hunk 10]
```diff
@@ -185,29 +251,33 @@ func (app *App) DeployService(service *Service) error {
...
 		if err != nil && !app.IgnoreDeployFailures {
 			return err
 		}
 	}
-
 	return nil
 }
 
-func (app *App) DisplayErrors() {
+func (app *App) handleErrors() {
```
Comment (svenwltr) [inline app.go]: Why did you change the name?

File: project_config.go [hunk 11]
```diff
@@ -1,13 +1,15 @@
 package main
```
Comment (svenwltr) [inline project_config.go]: Maybe also move the `ProjectConfig` to `settings`?

File: project_config.go [hunk 12]
```diff
@@ -42,3 +44,54 @@ func (c *ProjectConfig) WriteTo(filename string) error {
 
 	return ioutil.WriteFile(filename, data, 0644)
 }
+
+func (c *ProjectConfig) MergeConfig(localConfig *ProjectConfig) {
```
Comment (svenwltr) [inline project_config.go]: Looks like a candidate for an easy test.

File: app.go [hunk 13]
```diff
@@ -45,27 +51,64 @@ func (app *App) Run() error {
...
+	app.SleepInterval = *config.Settings.Sleep
+	app.IgnoreDeployFailures = *config.Settings.IgnoreDeployFailures
+	app.RetrySleep = *config.Settings.RetrySleep
+	app.RetryCount = *config.Settings.RetryCount
+	app.SkipShuffle = *config.Settings.SkipShuffle
+	app.SkipFetch = *config.Settings.SkipFetch
+	app.SkipDeploy = *config.Settings.SkipDeploy
+	app.WorkPath = workPath
+
+	if app.Kubectl == nil {
```
Comment (hendrik-rebuy) [inline app.go]: when is this not null?

Comment (svenwltr) [inline app.go]: No, it sets the value only, if it's not already set.

Comment (hendrik-rebuy) [inline app.go]: yes but when would it be set before?

Comment (svenwltr) [inline app.go]: In the tests, because otherwise it would deploy stuff on a real cluster.

File: app.go (hunk 10 above)
Comment (ptroc) [inline app.go]: Was needed for WorkPath clean up - not needed anymore, reverted the change

File: project_config.go (hunk 12 above)
Comment (ptroc) [inline project_config.go]: Done

**Comment by svenwltr** [comment]
What is with the skip flags? Do you want to put them back to the command line?

**Comment by ptroc** [comment]
@rebuy-de/prp-kubernetes-deployment
